File: app/src/main/java/org/dslul/openboard/inputmethod/latin/settings/AboutFragment.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 * modified
 * SPDX-License-Identifier: Apache-2.0 AND GPL-3.0-only
 */

package org.dslul.openboard.inputmethod.latin.settings;

import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.preference.Preference;

import org.dslul.openboard.inputmethod.latin.BuildConfig;
import org.dslul.openboard.inputmethod.latin.R;
import org.dslul.openboard.inputmethod.latin.utils.SpannableStringUtils;

/**
 * "About" sub screen.
 */
public final class AboutFragment extends SubScreenFragment {
    @Override
    public void onCreate(final Bundle icicle) {
        super.onCreate(icicle);
        addPreferencesFromResource(R.xml.prefs_screen_about);

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            // need to set icon tint because old android versions don't use the vector drawables
            for (int i = 0; i < getPreferenceScreen().getPreferenceCount(); i++) {
                final Preference p = getPreferenceScreen().getPreference(0);
                final Drawable icon = p.getIcon();
                if (icon != null)
                    DrawableCompat.setTint(icon, Color.WHITE);
            }
        }

        setupHiddenFeatures();
        setupVersionPref();
    }

    private void setupHiddenFeatures() {
        Preference hiddenFeaturesPreference = findPreference("hidden_features");
        hiddenFeaturesPreference.setOnPreferenceClickListener(preference -> {
            final String link = "<a href=\"https://developer.android.com/reference/android/content/Context#createDeviceProtectedStorageContext()\">"
                    + getString(R.string.hidden_features_text) + "</a>";
            final String message = requireContext().getString(R.string.hidden_features_message, link);
            final Spanned dialogMessage = SpannableStringUtils.fromHtml(message);

            final AlertDialog builder = new AlertDialog.Builder(requireContext())
                    .setIcon(R.drawable.ic_settings_about_hidden_features)
                    .setTitle(R.string.hidden_features_title)
                    .setMessage(dialogMessage)
                    .setPositiveButton(R.string.dialog_close, null)
                    .create();
            builder.show();
            ((TextView) builder.findViewById(android.R.id.message)).setMovementMethod(LinkMovementMethod.getInstance());
            return true;
        });
    }

    private void setupVersionPref() {
        Preference versionPreference = findPreference("pref_key_version");
        versionPreference.setSummary(BuildConfig.VERSION_NAME);
        if (BuildConfig.DEBUG) return;
        int[] count = new int[] {0};
        versionPreference.setOnPreferenceClickListener((pref) -> {
            if (getSharedPreferences().getBoolean(DebugSettings.PREF_SHOW_DEBUG_SETTINGS, false))
                return true;
            count[0]++;
            if (count[0] < 5) return true;
            getSharedPreferences().edit().putBoolean(DebugSettings.PREF_SHOW_DEBUG_SETTINGS, true).apply();
            Toast.makeText(requireContext(), R.string.prefs_debug_settings_enabled, Toast.LENGTH_LONG).show();
            return true;
        });
    }
}

File: app/src/main/java/org/dslul/openboard/inputmethod/latin/settings/AboutFragment.kt
/*
 * Copyright (C) 2014 The Android Open Source Project
 * modified
 * SPDX-License-Identifier: Apache-2.0 AND GPL-3.0-only
 */
package org.dslul.openboard.inputmethod.latin.settings

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import org.dslul.openboard.inputmethod.latin.BuildConfig
import org.dslul.openboard.inputmethod.latin.R
import org.dslul.openboard.inputmethod.latin.utils.Log
import org.dslul.openboard.inputmethod.latin.utils.SpannableStringUtils

/**
 * "About" sub screen.
 */
class AboutFragment : SubScreenFragment() {
    override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)
        addPreferencesFromResource(R.xml.prefs_screen_about)

        setupHiddenFeatures()
        setupVersionPref()
        findPreference<Preference>("log_reader")?.setOnPreferenceClickListener {
            val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
                .addCategory(Intent.CATEGORY_OPENABLE)
                .putExtra(Intent.EXTRA_TITLE,
                    requireContext().getString(R.string.english_ime_name)
                        .replace(" ", "_") + "_log_${System.currentTimeMillis()}.txt"
                )
                .setType("text/plain")
            logFilePicker.launch(intent)
            true
        }
    }

    private val logFilePicker = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode != Activity.RESULT_OK) return@registerForActivityResult
        val uri = result.data?.data ?: return@registerForActivityResult
        activity?.contentResolver?.openOutputStream(uri)?.use { os ->
            os.bufferedWriter().use { it.write(Log.getLog().joinToString("\n")) }
        }
    }

    private fun setupHiddenFeatures() {
        findPreference<Preference>("hidden_features")?.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val link = ("<a href=\"https://developer.android.com/reference/android/content/Context#createDeviceProtectedStorageContext()\">"
                            + getString(R.string.hidden_features_text) + "</a>")
                val message = requireContext().getString(R.string.hidden_features_message, link)
                val dialogMessage = SpannableStringUtils.fromHtml(message)
                val builder = AlertDialog.Builder(requireContext())
                        .setIcon(R.drawable.ic_settings_about_hidden_features)
                        .setTitle(R.string.hidden_features_title)
                        .setMessage(dialogMessage)
                        .setPositiveButton(R.string.dialog_close, null)
                        .create()
                builder.show()
                (builder.findViewById<View>(android.R.id.message) as TextView).movementMethod = LinkMovementMethod.getInstance()
                true
            }
    }

    private fun setupVersionPref() {
        val versionPreference = findPreference<Preference>("pref_key_version") ?: return
        versionPreference.summary = BuildConfig.VERSION_NAME
        if (BuildConfig.DEBUG) return
        var count = 0
        versionPreference.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                if (sharedPreferences.getBoolean(DebugSettings.PREF_SHOW_DEBUG_SETTINGS, false))
                    return@OnPreferenceClickListener true
                count++
                if (count < 5) return@OnPreferenceClickListener true
                sharedPreferences.edit().putBoolean(DebugSettings.PREF_SHOW_DEBUG_SETTINGS, true).apply()
                Toast.makeText(requireContext(), R.string.prefs_debug_settings_enabled, Toast.LENGTH_LONG).show()
                true
            }
    }
}